detectron2/data/datasets/coco.py

```diff
@@ -280,6 +287,117 @@ def file2id(folder_path, file_path):
+    categories = [
+        {"id": id, "name": name}
+        for id, name in enumerate(MetadataCatalog.get(dataset_name).thing_classes)
+    ]
+
+    logger.info("Converting dataset dicts into COCO format")
+    images = []
+    annotations = []
+
+    # just for logging purposes
+    _annotation_keys = Counter()
+
+    for image_dict in dataset_dicts:
+        image = {
+            "id": image_dict["image_id"],
+            "width": image_dict["width"],
+            "height": image_dict["height"],
+            "file_name": image_dict["file_name"],
+        }
+
+        images.append(image)
+
+        # deep-copying various annotations from the original format
+        # can be bbox, segmentation, keypoint, etc.
+        anns_per_image = deepcopy(image_dict["annotations"])
+        for annotation in anns_per_image:
+            # COCO requirement: linking annotations to images
+            annotation["id"] = len(annotations) + 1
+            annotation["image_id"] = image_dict["image_id"]
+
+            # COCO requirement: XYWH box format
+            bbox = annotation["bbox"]
+            bbox_mode = annotation["bbox_mode"]
+            bbox = BoxMode.convert(bbox, bbox_mode, BoxMode.XYWH_ABS)
+            del annotation["bbox_mode"]
+            # TODO: make BBOX_MODE serializable, otherwise remove it
+            annotation["bbox"] = bbox
+
+            # COCO requirement: instance area
+            if "segmentation" in annotation:
+                # Computing areas for instances by counting the pixels
+                segmentation = annotation["segmentation"]
+                # TODO: check segmentation type: RLE, BinaryMask or Polygon
+                polygons = PolygonMasks([segmentation])
+                area = polygons.area()[0]
+            else:
+                # Computing areas using bounding boxes
+                area = Boxes([bbox]).area()[0]
+            annotation["area"] = area
+
+            # Keeping track of fields present in instances
+            _annotation_keys.update(annotation.keys())
+
+            annotations.append(annotation)
+
+    logger.info(
+        "Conversion finished, " f"num images: {len(images)}, num annotations: {len(annotations)}"
+    )
+    logger.info(f"Annotation fields: {_annotation_keys}")
+    if len(_annotation_keys.most_common()) != len(_annotation_keys):
+        logger.warning(f"Annotation fields are not homogenous between instances")
+
+    info = {
+        "date_created": str(datetime.datetime.now()),
+        "description": "Automatically generated COCO json file for Detectron2.",
+    }
+    coco_dict = {
+        "info": info,
+        "images": images,
+        "annotations": annotations,
+        "categories": categories,
+        "licenses": None,
+    }
+    return coco_dict
+
+
+def convert_to_coco_json(dataset_name, allow_cached=True):
+    # TODO: The dataset or the conversion script *may* change,
+    # a checksum would be useful for validating the cached data
```

Comment:
Since the cached data is not a static resource (as you mentioned it may change), it would be better to generate it for every new training job. That shouldn't be very expensive I think.

Also, "/tmp" may not be a valid temporary directory on all systems, and even when it's valid it often takes RAM that won't be released. So the "output_dir" given to the evaluator seems to be a reasonable destination and in that case there is no need to clean the file in the end because a user expects to see artifacts dumped in that directory.

You can raise an error if the directory is not given. I think we always use evaluator with the directory at the moment.

Reply:
> That shouldn't be very expensive I think.

I don't think this is true, since we are populating a new dictionary with new elements, instead of just pointing to the already existing fields of the given dataset_dicts in a different format it can be very slow. 

Apropo caching: the cityscapes converter takes 10 mins for the training set, and 1 min for the val set. For personal usage, I've written a decorator that computes a hash based on the __repr__ of the arguments, concats the function's name, and caches the output in a pickle file of these computationally expensive pre-process steps. A quite ugly solution, but definitely essential for developing tools when one does not want to wait minutes to just check if their modifications works or not. @ppwwyyxx does it worth a PR?